### PR TITLE
Progress Report 2018 Part 2

### PR DESCRIPTION
**Pull request by CaptV0rt3x**
[Screenshot](https://i.imgur.com/iJTXSUp.png)

### CODE REVIEW

**Comment by Subv** [comment]
Some comments from Discord

```
there are a few commas that i feel shouldn't be there
first sentence of the GPU emulation section, for example
also, Puyo Puyo Tetris :smiley:
also, a nit: envydis has no documentation about the actual workings of each instruction
it just has the name, decoding and parameters for them
NEW MESSAGES
```

**Comment by DarkLordZach** [comment]
Unfortunately, we don't have XCI support yet. Someone else added the HFS container format (shiny i think) and that makes XCI easier, but we would still need to add the crypto. I'm looking into it, but right now XCI's will not boot. Sorry for the confusion.

**Comment by xavigisbeg** [comment]
Great post, it helped me understand the bigger picture of what I've been seeing following the PRs daily :+1:

Just nitpicking some typos:

Puyu Puyu -> Puyo Puyo 
disassmebler -> disassembler
nvidia -> Nvidia (?)
2d and 3d -> 2D and 3D (?)
til -> until (?)

Again, great job, I hope I can become skilled enough to start contributing a little myself :)

File: site/content/entry/yuzu-progress-report-2018-p1-2/index.md [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+modern GPUs it relies on a programmable pipeline. Basically this means that games write these little
+programs called shaders to control how triangles are drawn and how individual pixels are colored on
+screen.
+
+{{< imgs
+    "./debug.png|Behind the scenes!"
+>}}
+
+It wasn't just a matter of finding coordinates of triangles in memory, as these shader programs
+are run for every single vertex, triangle and pixel. We had to actually implement the shader programs,
```
Comment (jroweboy) [inline site/content/entry/yuzu-progress-report-2018-p1-2/index.md]: > implement 

lets make this a bit more specific

> We had to actually implement the shader programs to get more graphics output, meaning we needed to decompile each shader instruction. In case you are wondering, Nintendo Switch games are built with precompiled shader binaries, so we can't just run the original program; instead we have to analyze the shader instructions and figure out novel ways to get convert it back to a high level shader code again. Figuring out what each shader instruction means wasn't an easy task, as this...

File: site/content/entry/yuzu-progress-report-2018-p1-2/index.md [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+as the games were compacting with a compiled binary of instructions which controlled how shaders worked.
+Figuring out how this worked wasn't an easy task, as this was all nvidia's proprietary code. We worked
+a lot with [gdkchan](https://github.com/gdkchan), as he had made a bit of progress with this, and quickly
+learned how to decode GPU data based on nouveau and envytools. envytools and envydis (disassmebler)
+did the RE of how shaders work in Maxwell based GPUs, so we already know shader instructions and how
+they work generally.
```
Comment (jroweboy) [inline site/content/entry/yuzu-progress-report-2018-p1-2/index.md]: > we already know shader instructions and how they work generally

iirc we don't know how the work generally. isn't it more like 

> we already know how most of the disassemblies of the instructions, but not always what they do.

@Subv can you verify this ^

File: site/content/entry/yuzu-progress-report-2018-p1-2/index.md [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+
+{{< imgs
+    "./sonic_1.png"
+    "./sonic_2.png"
+		"./sonic_3.png"
+		"./sonic_4.png"
+		"./sonic_5.png"
+>}}
+
+For actual shader implementation in yuzu, we decided to base it on latest implementation done in Citra
```
Comment (jroweboy) [inline site/content/entry/yuzu-progress-report-2018-p1-2/index.md]: > latest implementation done

feels a little off. maybe this is cleaner?

> base it on a recent addition to Citra, a shader recompiler that converts the emulated 3DS shaders into GLSL so they can be run on the host GPU. Instead of running all...

File: site/content/entry/yuzu-progress-report-2018-p1-2/index.md [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+(Popularly known as GLvtx). This directly translates emulated shaders into OpenGL shaders. Basically
+instead of running all these shader programs (which have to run for thousands of vertices and millions
```
Comment (jroweboy) [inline site/content/entry/yuzu-progress-report-2018-p1-2/index.md]: > and millions of pixels

Drop this bit, thats the fragment shader, and citra's been doing the fragment shader on the GPU since 2015

File: site/content/entry/yuzu-progress-report-2018-p1-2/index.md [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+of pixels) on the CPU, we translate these shaders into OpenGL GLSL and upload the program directly as
+is to GPU. This is better as GPUs are designed to run these programs in parallel for thousands of vertices
+and millions of pixels, whereas CPUs are not. In Citra, we were doing this in software for a long time
```
Comment (jroweboy) [inline site/content/entry/yuzu-progress-report-2018-p1-2/index.md]: > we were doing this in software for a long time

Lets clarify that we weren't software rendered here. We were running fragment shaders on the GPU for a long time, and this is something other devs have gotten mad that we haven't been clear about when reporting the new feature.

> In Citra, we were running the fragment shaders on the GPU since at least 2015, but the vertex and geometry shaders were run on the CPU. The 3DS vertex and geometry shaders have some very hard to account for corner cases, and it took many years to make sure that our reverse engineering efforts were correct. That way we didn't have to waste effort working on something we couldn't be sure was possible. If you would like to read more about the feature in Citra, [check out the blog post about it](LINK TO POST)

pls add the link thx.

File: site/content/entry/yuzu-progress-report-2018-p1-2/index.md [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+before GLvtx mainly because of its simplicity and we were still learning.
+
+However this wasn't an option in case of yuzu, as Switch's GPU is modern and way too powerful for doing
```
Comment (jroweboy) [inline site/content/entry/yuzu-progress-report-2018-p1-2/index.md]: > However this wasn't an option in case of yuzu

weird tense shift from past to present.

> However this isnt't an option for yuzu, as the Switch's ...

File: site/content/entry/yuzu-progress-report-2018-p1-2/index.md [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+this in software. Around April 2018, we were able to get enough of shader instructions implemented to
+get "Puyu Puyu Tetris" show a bit of graphics. It wasn't much - SEGA logo and Tetris logo, and pretty
```
Comment (jroweboy) [inline site/content/entry/yuzu-progress-report-2018-p1-2/index.md]: > get "Puyu Puyu Tetris" show a bit of graphics

Puyo Puyo

also `to show` not just show

File: site/content/entry/yuzu-progress-report-2018-p1-2/index.md [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+much hung after that. Further, we were able to get a few other simple 2d games like "Cave Story" and
```
Comment (jroweboy) [inline site/content/entry/yuzu-progress-report-2018-p1-2/index.md]: `Futhermore,`

File: site/content/entry/yuzu-progress-report-2018-p1-2/index.md [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+"The Binding of Issac" booting as well.
+
+{{< imgs
+    "./splatoon2.mp4|Splatoon 2 - Then"
+    "./splatoon2_work.png|Splatoon 2 - Now"
+>}}
+
+Another new feature which we implemented was shader `constbuffer` support by Subv. We have shader programs
+that we want to reuse for a bunch of different triangles or features or UI things or 3d models. It isn't
```
Comment (jroweboy) [inline site/content/entry/yuzu-progress-report-2018-p1-2/index.md]: > different triangles or features or UI things or 3d models

little weird. How about

> different triangles for features like the UI or 3d models.

File: site/content/entry/yuzu-progress-report-2018-p1-2/index.md [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+efficient to have individual shaders for every single thing, instead we can have a couple of shader
+programs that can be reused for everything. So, shader `constbuffer` support is the equivalent of OpenGL
+`uniform` and `uniform buffer objects (UBO)`. Uniforms in general are a way to provide constant data
+to shaders and can also be used to reconfigure the shader as well.
+
+Initial rendering support and blending support were done by bunnei. Blending is used for alpha transparency
+- translating Switch GPU register writes to OpenGL calls. It fixed a bug in "Puyu Puyu Tetris", where
```
Comment (jroweboy) [inline site/content/entry/yuzu-progress-report-2018-p1-2/index.md]: Puyu Puyu 

pls Puyo Puyo

File: site/content/entry/yuzu-progress-report-2018-p1-2/index.md [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+you couldn't save the icon in top left corner. Initial texture support was done by Subv and mainly
```
Comment (jroweboy) [inline site/content/entry/yuzu-progress-report-2018-p1-2/index.md]: > where you couldn't save the icon in top left corner

the actual bug was it had a weird box around it

> where the `Saving...` icon in the top left corner had a weird box around it.

File: site/content/entry/yuzu-progress-report-2018-p1-2/index.md [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+involved reading a texture from memory, decoding it, and uploading it to OpenGL.
+
+Coming to rasterizer cache, its computationally expensive to upload texture from Switch memory every
```
Comment (jroweboy) [inline site/content/entry/yuzu-progress-report-2018-p1-2/index.md]: > to upload texture 

just a bit of clarification

> to upload a texture from the emulated Switch memory to the host GPU every time it's used.

File: site/content/entry/yuzu-progress-report-2018-p1-2/index.md [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+time its used. The memory needs to be copied from CPU space to GPU space, the texture needs to be
```
Comment (jroweboy) [inline site/content/entry/yuzu-progress-report-2018-p1-2/index.md]: > copied from CPU space to GPU space, 

Lets switch this around because we will talk more about the copying after this, and don't talk about swizzling at all.

> The texture will need to be decoded and unswizzled before its uploaded, and then the memory will need to be copied from the CPU RAM into the GPU RAM.

File: site/content/entry/yuzu-progress-report-2018-p1-2/index.md [hunk 14]
```diff
@@ -0,0 +1,233 @@
...
+decoded and unswizzled. It's much more efficient to just keep texture in the GPU memory, similar to
+how it works on hardware with normal PC games, but is a little bit trickier for emulation. This is
+because we don't know when the game might change the texture or modify it. bunnei did most of the work
+to essentially cache these textures, so when a texture is uploaded to OpenGL it is saved in GPU memory
```
Comment (jroweboy) [inline site/content/entry/yuzu-progress-report-2018-p1-2/index.md]: > to essentially cache these textures

split infinitive reeeeee

just remove essentially

File: site/content/entry/yuzu-progress-report-2018-p1-2/index.md [hunk 15]
```diff
@@ -0,0 +1,233 @@
...
+and we keep track of it. We detect when the texture is changed by the game and refresh it as necessary.
```
Comment (jroweboy) [inline site/content/entry/yuzu-progress-report-2018-p1-2/index.md]: > We detect when the texture is changed by the game 

Little bit of detail on this

> When the CPU or our emulated Switch kernel reads or writes to the memory address that the texture was uploaded to, we check to see what we need to do, and reload the texture only if necessary.

File: site/content/entry/yuzu-progress-report-2018-p1-2/index.md [hunk 16]
```diff
@@ -0,0 +1,233 @@
...
+This also applies to framebuffers because in some cases, games can use them as textures. If this caching
+wasn't existent, we'd essentially have to upload every texture to OpenGL memory (GPU space) on every
```
Comment (jroweboy) [inline site/content/entry/yuzu-progress-report-2018-p1-2/index.md]: > GPU space

clarify it as host GPU space

File: site/content/entry/yuzu-progress-report-2018-p1-2/index.md [hunk 17]
```diff
@@ -0,0 +1,233 @@
...
+draw, copy the final framebuffer back to Switch memory (CPU space) so that the game could potentially
```
Comment (jroweboy) [inline site/content/entry/yuzu-progress-report-2018-p1-2/index.md]: > back to Switch memory 

clarify this as `emulated Switch memory`

File: site/content/entry/yuzu-progress-report-2018-p1-2/index.md [hunk 18]
```diff
@@ -0,0 +1,233 @@
...
+>}}
+
+{{< imgs
+		"./SMO_debug.png|Debugging"
+		"./SMO_work.png| Proper render"
+		"./SMO_2_work.png| Proper render"
+>}}
+
+We had a couple of fixes to our rasterizer cache mainly because it was based on Citra's rasterizer cache.
+As 3DS didn't require any compressed texture formats in OpenGL, Citra didn't implement it in its cache.
```
Comment (jroweboy) [inline site/content/entry/yuzu-progress-report-2018-p1-2/index.md]: > 3DS didn't require any compressed texture formats in OpenGL

The 3DS does use compressed formats like ETC1, but most desktop gpus don't support that so we decoded them 

> Because the compressed texture formats that the 3DS uses are not commonly supported by desktop GPUs, Citra's rasterizer cache decodes any compressed textures before uploading and caching them. But since the Switch GPU supports many of the same texture formats that desktop OpenGL does, we sped things up by uploading them without decoding them. At the start, this support was just hacked into Citra's cache, which didn't end up working correctly all the time. One example was the squid texture in splatoon 2...

File: site/content/entry/yuzu-progress-report-2018-p1-2/index.md [hunk 19]
```diff
@@ -0,0 +1,233 @@
...
+But as the Switch uses them, we had to implement it. Because of not implementing this in the rasterizer
+cache, there were lot of scenarios which weren't well supported by it. One of them is the squid textures
+in "Splatoon 2", which was identified and fixed by Subv. We also implemented texture color swizzling,
+a way to swap color components, which is used by "Splatoon 2" to swap colors for intro background
+texture.So, before we implemented this the colors were all wrong. We fixed YUV2 video playback in our
```
Comment (jroweboy) [inline site/content/entry/yuzu-progress-report-2018-p1-2/index.md]: > .So, befo

Missing space after period

File: site/content/entry/yuzu-progress-report-2018-p1-2/index.md [hunk 20]
```diff
@@ -0,0 +1,233 @@
...
+shader support, which fixed the previously busted "Sonic Mania" intro. We had to implement a field
+parameter called component masks in TEX/TEXS shader instructions, to properly decode YUV2 videos.
+
+We also implemented texture wrap modes, which gives the ability to specify whether a texture on a
```
Comment (jroweboy) [inline site/content/entry/yuzu-progress-report-2018-p1-2/index.md]: Put the texture wrap mode sentence in the previous paragraph and start a new one with `There are lots of ...`

File: site/content/entry/yuzu-progress-report-2018-p1-2/index.md [hunk 21]
```diff
@@ -0,0 +1,233 @@
...
+
+Apart from the above mentioned stuff, there have been many more modifications made to yuzu's GPU HLE.
+These bug fixes or modifications were done on a per-game experimentation basis and as we progress further
+we will continue to fix our implementations and improve the accuracy of emulation in yuzu. All of the
+progress we have made til now is thanks to the efforts of yuzu's contributors and the good people from
+Switch hacking communities.
+
+## Virtual File System (VFS) by [DarkLordZach](https://github.com/DarkLordZach) ([here](https://github.com/yuzu-emu/yuzu/pull/679))
+
+A Virtual File System (VFS) is an abstraction layer on top of a more concrete file system. The purpose
```
Comment (jroweboy) [inline site/content/entry/yuzu-progress-report-2018-p1-2/index.md]: > abstraction layer on top of a more concrete file system

Eh. just say it like this

> abstraction layer which allows us to hide the details of where the actual files are stored. The purpose of a VFS is to enable the emulated Switch file system to read and write to many different types of backends, without changing anything in the emulated FS code. In layman's terms, the game will still think its reading and writing to a file on the Switch, but in reality it could be reading from a zip file, or from a custom mod that the user adds. This makes further support for Updates...

File: site/content/entry/yuzu-progress-report-2018-p1-2/index.md [hunk 22]
```diff
@@ -0,0 +1,233 @@
...
+
+DarkLordZach has been working very actively on various bug fixes, features and has also been lending
+a hand in testing games too. His latest contribution comes in the form of touch-screen support for
+yuzu. With this feature, now yuzu emulates mouse clicks to touch inputs and if you happen to have a
+physical touch screen those would convert into inputs as well.
+
+## Miscellaneous
+
+Apart from these improvements, we have had multiple PRs which fixed bugs and deadlocks in various games.
+We got graphics output in Super Mario Odyssey, fixed - saves not working, got bunch of 4.x games
```
Comment (jroweboy) [inline site/content/entry/yuzu-progress-report-2018-p1-2/index.md]: > fixed - saves not working

this one is negative (saves NOT working) and the rest are positive (got game booting) change it to

> fixed save related issues

Comment (jroweboy) [reply site/content/entry/yuzu-progress-report-2018-p1-2/index.md]: > got bunch of 4.x games

different from the others. all the others `past tense verb + noun`

> launched many new games to the title screen

File: site/content/entry/yuzu-progress-report-2018-p1-2/index.md [hunk 23]
```diff
@@ -0,0 +1,233 @@
...
+
+{{< imgs
+    "./sonic_glitch.png|Sonic Mania - Then"
+    "./sonicmania.mp4|Sonic Mania - Now"
+>}}
+
+With the help of these projects we made good progress in GPU emulation and got games to submit drawing
+calls, not just configuring the hardware but actually writing to registers, indicating that they were
+ready to draw triangles on the screen. Even the simplest 2d games are not just 2d anymore. As the Switch
+uses a modern GPU, there is no hardware for drawing 2d anymore. We draw 2 triangles to represent a
```
[the remark this answers is not in the document]
Comment (Subv) [reply site/content/entry/yuzu-progress-report-2018-p1-2/index.md]: Sure that's accurate

File: site/content/entry/yuzu-progress-report-2018-p1-2/index.md (hunk 2 above)
Comment (Subv) [reply site/content/entry/yuzu-progress-report-2018-p1-2/index.md]: yeah, we can know the format and bit layout of the instruction because envydis is awesome, but for the most part we're left in the dark about what their parameters mean or they _actually_ do (gratuitous example, what does the `isberd` instruction do?)

File: site/content/entry/yuzu-progress-report-2018-p1-2/index.md (hunk 10 above)
Comment (Subv) [reply site/content/entry/yuzu-progress-report-2018-p1-2/index.md]: Puyo
